### PR TITLE
BUG: fix kwarg handling in assert_warn [skip cirrus][skip azp]

### LINKED ISSUE #25989: BUG: `assert_warns` ignores `kwargs` unless `args` is passed
### Describe the issue:

Some tests in SciPy (scipy/scipy#20031) were written under the assumption that `assert_warns` accepts a `match` parameter like `pytest.warns`. Unfortunately, `assert_warns` accepts the parameter but ignores it, so these tests passed even though the intended warning was not emitted.

In these cases, I think `assert_warns` should have raised an error when attempting to pass `**kwargs` to a user-provided callable, but this isn't attempted unless `*args` is passed. Should there be an error if `**kwargs` is passed without `*args`?

### Reproduce the code example:

```python
from numpy.testing import assert_warns
import warnings

with assert_warns(UserWarning, match="A"):
    warnings.warn("B", UserWarning)
```

### Error message:

_No response_

### Python and NumPy Versions:

1.25.2
3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]

### Runtime Environment:

_No response_

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by mattip**
Fixes #25989

### CODE REVIEW

**Comment by mattip** [comment]
I skipped azure and cirrus since the pure-python code should be tested well enough by github actions. This is a bugfix, is it important enough for a release note?

File: numpy/testing/tests/test_utils.py [hunk 1]
```diff
@@ -1035,6 +1035,19 @@ def no_warnings():
...
+            warnings.warn("yo")
+            return a + b
+
+        assert assert_warns(UserWarning, f, b=20) == 20
+
+        with pytest.raises(RuntimeError) as exc:
+            # assert_warns cannot do regexp matching, use pytest.warn
+            with assert_warns(UserWarning, match="A"):
+                warnings.warn("B", UserWarning)
+            assert "assert_warns" in str(exc)
```
Comment (mdhaber) [inline numpy/testing/tests/test_utils.py:1049]: Why not

```suggestion
        with pytest.raises(RuntimeError, match="assert_warns called...") as exc:
            # assert_warns cannot do regexp matching, use pytest.warn
            with assert_warns(UserWarning, match="A"):
                warnings.warn("B", UserWarning)
```

File: numpy/testing/_private/utils.py [hunk 2]
```diff
@@ -1949,8 +1949,10 @@ def assert_warns(warning_class, *args, **kwargs):
     >>> ret = np.testing.assert_warns(DeprecationWarning, deprecated_func, 4)
     >>> assert ret == 16
     """
-    if not args:
+    if not args and not kwargs:
         return _assert_warns_context(warning_class)
+    elif len(args) < 1:
+        raise RuntimeError("assert_warns called without args")
```
Comment (mdhaber) [inline numpy/testing/_private/utils.py:1955]: Other libraries might have done the same thing we did in SciPy, thinking that `assert_warns` processed `match` like `pytest.warns`. If so, they may be confused when they start getting this error. Not sure what the best thing to do is, but one possibility is to check whether `'match' in kwargs`, and it so give a little more information.

**Comment by mdhaber** [comment]
> This is a bugfix, is it important enough for a release note?

Typically, no. The only thought is if it would [reduce confusion](https://github.com/numpy/numpy/pull/25999#pullrequestreview-1930104528) and possibly save the hassle of bug reports. (But I think a more detailed error message would be more effective at that.)

File: numpy/testing/_private/utils.py (hunk 2 above)
Comment (mattip) [reply numpy/testing/_private/utils.py]: I improved the error handling and the test

**Comment by mdhaber** [comment]
Thanks, I think that will save users and possibly NumPy maintainers some trouble later. Just wondered why not use the `match` parameter of `pytest.raises`? If you would prefer not to, I think this would resolve the main concern of gh-25989.

**Comment by mattip** [comment]
> manually checking the error message string

Do you mean in the test? When a test with a regexp fails, it always takes me soooo long to figure out what might be going wrong. This way I get a clearer indication, at the cost of one more line of code.

**Pull request merged**